### PR TITLE
hlw8012: fix constants for BL0937

### PR DESCRIPTION
**Pull request by ianchi**
With the current implementation it is impossible to calibrate all measurements at the same time for the BL0937 sensor as it has different constants factors for converting pulses.

If you set `voltage_divider` and `current_resistor` to correctly calibrate *voltage* and *current*, you get a reading for *power* that is completely off.

Add a configuration option `sensor_model` to handle constants on per sensor model basis.

While at it, avoid computing the same factor on every update and move to setup logic to do it only once.

## Types of changes

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes BL0937 sensor readings

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1282
## Test Environment

- [ ] ESP32
- [x] ESP8266

## Example entry for `config.yaml`:

```yaml
sensor:
  - platform: hlw8012
    sel_pin:
      number: GPIO12
      inverted: true
    cf_pin: GPIO4
    cf1_pin: GPIO5
    sensor_model: BL0937
    voltage_divider: 1685

    current:
      name: Current
    voltage:
      name: Voltage
    power:
      name: Power
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Review by jesserockz — COMMENTED** [review]
I cant help but think there is a reason for the `/ 1000000.0f` in the final calculations. Perhaps precision is lost if the multiplier is a smaller number/more decimals before multiplying the `cf_hz`/`cf1_hz`?
(If I am spitting out garbage, just let me know.)

The only other comment I have is probably, use the existing `CONF_MODEL`. You don't need to make a new one and call everything `...SensorModel` when `Model` will do fine.

**Comment by OttoWinter** [comment]
@jesserockz That depends a bit on the scale of those values, but iirc those constants were just there for easier code (because iirc the datasheet uses those units). If anything, I _think_ it should be more accurate now, because IEEE754 floats have higher precision closer to 0.

**Comment by ianchi** [comment]
Fixed name of config option to `model`.
Changed docs PR accordingly.

File: esphome/components/hlw8012/sensor.py [hunk 1]
```diff
@@ -62,6 +71,7 @@
         ),
         cv.Optional(CONF_CURRENT_RESISTOR, default=0.001): cv.resistance,
         cv.Optional(CONF_VOLTAGE_DIVIDER, default=2351): cv.positive_float,
+        cv.Optional(CONF_MODEL, default="HLW8012"): cv.one_of(*MODELS, upper=True),
```
Comment (jesserockz) [inline esphome/components/hlw8012/sensor.py]:
```suggestion
        cv.Optional(CONF_MODEL, default="HLW8012"): cv.enum(MODELS, upper=True),
```

File: esphome/components/hlw8012/sensor.py [hunk 2]
```diff
@@ -99,3 +109,4 @@ async def to_code(config):
     cg.add(var.set_voltage_divider(config[CONF_VOLTAGE_DIVIDER]))
     cg.add(var.set_change_mode_every(config[CONF_CHANGE_MODE_EVERY]))
     cg.add(var.set_initial_mode(INITIAL_MODES[config[CONF_INITIAL_MODE]]))
+    cg.add(var.set_sensor_model(MODELS[config[CONF_MODEL]]))
```
Comment (jesserockz) [inline esphome/components/hlw8012/sensor.py]:
```suggestion
    cg.add(var.set_sensor_model(config[CONF_MODEL]))
```

**Review by jesserockz — COMMENTED** [review]
Just 2 suggestions below to use the `cv.enum` helper

File: esphome/const.py [hunk 3]
```diff
@@ -522,6 +522,7 @@
 CONF_SENSING_PIN = "sensing_pin"
 CONF_SENSOR = "sensor"
 CONF_SENSOR_ID = "sensor_id"
+CONF_SENSOR_MODEL = "sensor_model"
```
Comment (jesserockz) [inline esphome/const.py]:
```suggestion
```

**Review by jesserockz — APPROVED** [review]

**Pull request merged**